File: ios/Classes/SocureFlutterPlugin.swift
import Flutter
import UIKit
import SocureDocV

public class SocureFlutterPlugin: NSObject, FlutterPlugin {

  private var rootViewController: UIViewController? {
    return UIApplication.shared.keyWindow?.rootViewController
  }

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "socure_flutter", binaryMessenger: registrar.messenger())
    let instance = SocureFlutterPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)
    case "launchDocV":
      launchDocV(call: call, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func launchDocV(call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any] else {
      result(FlutterError(
        code: "INVALID_ARGUMENTS",
        message: "Invalid arguments provided",
        details: nil
      ))
      return
    }

    guard let sdkKey = args["sdkKey"] as? String, !sdkKey.isEmpty else {
      result(FlutterError(
        code: "INVALID_KEY",
        message: "SDK key is required",
        details: nil
      ))
      return
    }

    guard let transactionToken = args["transactionToken"] as? String, !transactionToken.isEmpty else {
      result(FlutterError(
        code: "INVALID_TOKEN",
        message: "Transaction token is required",
        details: nil
      ))
      return
    }

    guard let viewController = rootViewController else {
      result(FlutterError(
        code: "NO_VIEW_CONTROLLER",
        message: "No view controller available to present DocV",
        details: nil
      ))
      return
    }

    let useSocureGov = args["useSocureGov"] as? Bool ?? false

    // Create Socure DocV options
    let options = SocureDocVOptions(
      publicKey: sdkKey,
      docvTransactionToken: transactionToken,
      presentingViewController: viewController,
      useSocureGov: useSocureGov
    )

    // Launch Socure DocV SDK
    SocureDocVSDK.launch(options) { sdkResult in
      switch sdkResult {
      case .success(let success):
        let resultDict: [String: Any] = [
          "success": true,
          "deviceSessionToken": success.deviceSessionToken ?? ""
        ]
        result(resultDict)

      case .failure(let failure):
        let errorCode = self.mapErrorToCode(failure.error)
        var resultDict: [String: Any] = [
          "success": false,
          "errorCode": errorCode,
          "errorMessage": String(describing: failure.error)
        ]

        // Include deviceSessionToken if available
        if let sessionToken = failure.deviceSessionToken {
          resultDict["deviceSessionToken"] = sessionToken
        }

        result(resultDict)
      }
    }
  }

  private func mapErrorToCode(_ error: SocureDocVError) -> String {
    switch error {
    case .invalidPublicKey:
      return "INVALID_KEY"
    case .invalidDocvTransactionToken:
      return "INVALID_TOKEN"
    case .noInternetConnection:
      return "NETWORK_ERROR"
    case .userCanceled:
      return "USER_CANCELED"
    case .cameraPermissionDeclined:
      return "CAMERA_PERMISSION_DENIED"
    case .sessionExpired:
      return "SESSION_EXPIRED"
    case .sessionInitiationFailure:
      return "INITIALIZATION_ERROR"
    case .documentUploadFailure:
      return "CAPTURE_ERROR"
    case .consentDeclined:
      return "CONSENT_DECLINED"
    case .unknown:
      return "UNKNOWN"
    @unknown default:
      return "UNKNOWN"
    }
  }
}

File: android/src/main/kotlin/com/aniketlabs/socure_flutter/SocureFlutterPlugin.kt
package com.aniketlabs.socure_flutter

import android.app.Activity
import android.content.Context
import android.content.Intent
import com.socure.docv.capturesdk.api.SocureSdk
import com.socure.docv.capturesdk.api.SocureDocVContext
import com.socure.docv.capturesdk.api.SocureDocVError
import com.socure.docv.capturesdk.common.utils.SocureDocVFailure
import com.socure.docv.capturesdk.common.utils.SocureDocVSuccess
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry

/** SocureFlutterPlugin */
class SocureFlutterPlugin :
    FlutterPlugin,
    MethodCallHandler,
    ActivityAware,
    PluginRegistry.ActivityResultListener {

    private lateinit var channel: MethodChannel
    private var activity: Activity? = null
    private var context: Context? = null
    private var pendingResult: Result? = null

    companion object {
        private const val REQUEST_CODE_DOCV = 1001
    }

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "socure_flutter")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
    }

    override fun onMethodCall(
        call: MethodCall,
        result: Result
    ) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }
            "launchDocV" -> {
                launchDocV(call, result)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun launchDocV(call: MethodCall, result: Result) {
        val currentActivity = activity
        if (currentActivity == null) {
            result.error(
                "NO_ACTIVITY",
                "Activity not available",
                null
            )
            return
        }

        // Check if there's already a pending result
        if (pendingResult != null) {
            result.error(
                "ALREADY_ACTIVE",
                "A DocV session is already active",
                null
            )
            return
        }

        try {
            // Extract parameters
            val sdkKey = call.argument<String>("sdkKey")
            val transactionToken = call.argument<String>("transactionToken")
            val useSocureGov = call.argument<Boolean>("useSocureGov") ?: false

            if (sdkKey.isNullOrEmpty()) {
                result.error("INVALID_KEY", "SDK key is required", null)
                return
            }

            if (transactionToken.isNullOrEmpty()) {
                result.error("INVALID_TOKEN", "Transaction token is required", null)
                return
            }

            // Store the result callback for later
            pendingResult = result

            // Build Socure DocV context (based on official Socure docs)
            val docVContext = SocureDocVContext(
                transactionToken,
                sdkKey,
                useSocureGov
            )

            // Get intent from Socure SDK
            val intent = SocureSdk.getIntent(currentActivity, docVContext)

            // Launch the DocV activity
            currentActivity.startActivityForResult(intent, REQUEST_CODE_DOCV)

        } catch (e: Exception) {
            pendingResult = null
            result.error(
                "INITIALIZATION_ERROR",
                "Failed to launch DocV: ${e.message}",
                null
            )
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?): Boolean {
        if (requestCode == REQUEST_CODE_DOCV) {
            val result = pendingResult
            pendingResult = null

            if (result == null) {
                return true
            }

            if (data == null) {
                result.error(
                    "NO_DATA",
                    "No data returned from DocV SDK",
                    null
                )
                return true
            }

            // Process result using Socure SDK
            SocureSdk.getResult(data) { sdkResult ->
                when (sdkResult) {
                    is SocureDocVSuccess -> {
                        val resultMap = mapOf(
                            "success" to true,
                            "deviceSessionToken" to sdkResult.deviceSessionToken
                        )
                        result.success(resultMap)
                    }
                    is SocureDocVFailure -> {
                        val errorCode = mapErrorToCode(sdkResult.error)
                        val resultMap = mapOf(
                            "success" to false,
                            "errorCode" to errorCode,
                            "errorMessage" to sdkResult.error.toString(),
                            "deviceSessionToken" to sdkResult.deviceSessionToken
                        )
                        result.success(resultMap)
                    }
                }
            }
            return true
        }
        return false
    }

    private fun mapErrorToCode(error: SocureDocVError): String {
        return when (error) {
            SocureDocVError.INVALID_PUBLIC_KEY -> "INVALID_KEY"
            SocureDocVError.INVALID_DOCV_TRANSACTION_TOKEN -> "INVALID_TOKEN"
            SocureDocVError.NO_INTERNET_CONNECTION -> "NETWORK_ERROR"
            SocureDocVError.USER_CANCELED -> "USER_CANCELED"
            SocureDocVError.CAMERA_PERMISSION_DECLINED -> "CAMERA_PERMISSION_DENIED"
            SocureDocVError.SESSION_EXPIRED -> "SESSION_EXPIRED"
            SocureDocVError.SESSION_INITIATION_FAILURE -> "INITIALIZATION_ERROR"
            SocureDocVError.DOCUMENT_UPLOAD_FAILURE -> "CAPTURE_ERROR"
            SocureDocVError.CONSENT_DECLINED -> "CONSENT_DECLINED"
            SocureDocVError.UNKNOWN -> "UNKNOWN"
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        context = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        activity = binding.activity
        binding.addActivityResultListener(this)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        activity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        activity = binding.activity
        binding.addActivityResultListener(this)
    }

    override fun onDetachedFromActivity() {
        activity = null
    }
}
